python/utils/PyTypes.h

```diff
@@ -0,0 +1,168 @@
+  Complex(std::complex<double> value)
+      : Complex((double)value.real(), (double)value.imag()) {}
+
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  Complex(std::complex<float> value)
+      : Complex((double)value.real(), (double)value.imag()) {}
+
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  operator std::complex<double>() {
+    auto value = PyComplex_AsCComplex(m_ptr);
+    return std::complex<double>(value.real, value.imag);
+  }
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  operator std::complex<float>() {
+    auto value = PyComplex_AsCComplex(m_ptr);
+    return std::complex<float>(value.real, value.imag);
+  }
+
+  static bool isComplex_(PyObject *o) {
+    if (PyComplex_Check(o)) {
+      return true;
+    }
+    PyTypeObject *type = Py_TYPE(o);
+    std::string name = std::string(type->tp_name);
+    if (name == "numpy.complex64" || name == "numpy.complex128") {
+      return true;
+    }
+    return false;
+  }
+
+  static PyObject *convert_(PyObject *o) {
+    PyObject *ret = nullptr;
+    if (isComplex_(o)) {
+      double real = PyComplex_RealAsDouble(o);
+      double imag = PyComplex_ImagAsDouble(o);
+      ret = PyComplex_FromDoubles(real, imag);
+    } else {
+      py::set_error(PyExc_TypeError, "Unexpected type");
+    }
+    return ret;
+  }
+};
+
+/// Extended python float object.
+///
+/// Includes `float`, `numpy.float64`, `numpy.float32`.
+class Float : public py::object {
+public:
+  PYBIND11_OBJECT_CVT(Float, object, isFloat_, convert_)
+
+  // Allow implicit conversion from float/double:
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  Float(float value) : object(PyFloat_FromDouble((double)value), stolen_t{}) {
+    if (!m_ptr) {
+      py::pybind11_fail("Could not allocate float object!");
+    }
+  }
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  Float(double value = .0)
+      : object(PyFloat_FromDouble((double)value), stolen_t{}) {
+    if (!m_ptr) {
+      py::pybind11_fail("Could not allocate float object!");
+    }
+  }
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  operator float() const { return (float)PyFloat_AsDouble(m_ptr); }
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  operator double() const { return (double)PyFloat_AsDouble(m_ptr); }
+
+  static bool isFloat_(PyObject *o) {
+    if (PyFloat_Check(o)) {
+      return true;
+    }
+    PyTypeObject *type = Py_TYPE(o);
+    std::string name = std::string(type->tp_name);
+    if (name == "numpy.float32" || name == "numpy.float64") {
+      return true;
+    }
+    return false;
+  }
+
+  static PyObject *convert_(PyObject *o) {
+    PyObject *ret = nullptr;
+    if (isFloat_(o)) {
+      ret = PyFloat_FromDouble(PyFloat_AsDouble(o));
+    } else {
+      py::set_error(PyExc_TypeError, "Unexpected type");
+    }
+    return ret;
+  }
+};
+
+template <typename T>
+inline std::string typeName() {
+  return {typeid(T).name()};
+}
+template <>
+inline std::string typeName<py_ext::Float>() {
+  return "float";
```

Comment:
Curious: do these functions have to return `std::string` objects? Or can they just as easily return `char const *` values that may (or may not) need to be copied to the heap in the calling code?

Reply:
They definitely can return char*, will change!